File: modules/admin/admin-ui/src/main/resources/web/admin/common/js/data/Property.ts
module api.data {

    import BinaryReference = api.util.BinaryReference;
    import Reference = api.util.Reference;
    import GeoPoint = api.util.GeoPoint;
    import LocalTime = api.util.LocalTime;
    import DateTime = api.util.DateTime;
    import LocalDateTime = api.util.LocalDateTime;
    import LocalDate = api.util.LocalDate;
    
    /**
     * A Property has a [[name]] and a [[value]],
     * but also:
     * *  an [[index]], since it's a part of an [[array]]
     * *  a [[parent]], since it's also a part of a [[PropertySet]]
     *
     * A Property is mutable, both it's [[index]] and [[value]] can change.
     */
    export class Property implements api.Equitable {

        public static debug: boolean = false;

        private parent: PropertySet;

        private array: PropertyArray;

        private name: string;

        private index: number;

        private value: Value;

        private propertyIndexChangedListeners: {(event: PropertyIndexChangedEvent):void}[] = [];

        private propertyValueChangedListeners: {(event: PropertyValueChangedEvent):void}[] = [];

        constructor(builder: PropertyBuilder) {
            api.util.assertNotNull(builder.array, "array of a Property cannot be null");
            api.util.assertNotNull(builder.name, "name of a Property cannot be null");
            api.util.assertNotNull(builder.index, "index of a Property cannot be null");
            api.util.assertNotNull(builder.value, "value of a Property cannot be null");

            this.array = builder.array;
            this.parent = builder.array.getParent();
            this.name = builder.name;
            this.index = builder.index;
            this.value = builder.value;

            if (this.value.getType().equals(ValueTypes.DATA) && this.value.isNotNull()) {
                var valuePropertySet = this.value.getPropertySet();
                valuePropertySet.setContainerProperty(this);
            }
        }

        /**
         * Change the index.
         *
         * A [[PropertyIndexChangedEvent]] will be notified to listeners if the index really changed.
         * @param newIndex
         */
        setIndex(newIndex: number) {
            var oldIndex = newIndex;
            this.index = newIndex;

            if (oldIndex != newIndex) {
                this.notifyPropertyIndexChangedEvent(oldIndex, newIndex);
            }
        }

        /**
         * Change the value.
         *
         * A [[PropertyValueChangedEvent]] will be notified to listeners if the value really changed.
         * @param value
         */
        setValue(value: Value) {
            api.util.assertNotNull(value, "value of a Property cannot be null");
            var oldValue = this.value;
            this.value = value;

            // Register listeners on PropertySet
            if (this.value.getType().equals(ValueTypes.DATA) && this.value.isNotNull()) {
                var propertySet = this.value.getPropertySet();
                propertySet.setContainerProperty(this);
                this.array.registerPropertySetListeners(propertySet);
            }

            // Unregister listeners on PropertySet from oldValue
            if (oldValue.getType().equals(ValueTypes.DATA) && oldValue.isNotNull()) {
                var removedPropertySet = oldValue.getPropertySet();
                removedPropertySet.setContainerProperty(null);
                this.array.unregisterPropertySetListeners(removedPropertySet);
            }

            if (!value.equals(oldValue)) {
                this.notifyPropertyValueChangedEvent(oldValue, value);
            }
        }

        convertValueType(type: ValueType) {
            this.array.convertValues(type);
        }

        /**
         * Detach this Property from it's array and parent. Should be called when removed from the array.
         */
        detach() {
            this.array = null;
            this.parent = null;
            this.propertyIndexChangedListeners = [];
            this.propertyValueChangedListeners = [];
        }

        getParent(): PropertySet {
            return this.parent;
        }

        hasParentProperty(): boolean {
            return !!this.getParentProperty();
        }

        getParentProperty(): Property {
            return this.parent.getProperty();
        }

        getPath(): PropertyPath {
            if (this.hasParentProperty()) {
                return PropertyPath.fromParent(this.getParentProperty().getPath(), new PropertyPathElement(this.name, this.index))
            }
            else {
                return PropertyPath.fromPathElement(new PropertyPathElement(this.name, this.index));
            }
        }

        getName(): string {
            return this.name;
        }

        getIndex(): number {
            return this.index;
        }

        getType(): ValueType {
            return this.value.getType();
        }

        getValue(): Value {
            return this.value;
        }

        hasNullValue(): boolean {
            return this.value.isNull();
        }

        hasNonNullValue(): boolean {
            return !this.value.isNull();
        }

        getPropertySet(): PropertySet {
            return this.value.getPropertySet();
        }

        getString(): string {
            return this.value.getString();
        }

        getLong(): number {
            return this.value.getLong();
        }

        getDouble(): number {
            return this.value.getDouble();
        }

        getBoolean(): boolean {
            return this.value.getBoolean();
        }

        getDateTime(): DateTime {
            return this.value.getDateTime();
        }

        getLocalDate(): LocalDate {
            return this.value.getLocalDate();
        }

        getLocalDateTime(): LocalDateTime {
            return this.value.getLocalDateTime();
        }

        getLocalTime(): LocalTime {
            return this.value.getLocalTime();
        }

        getGeoPoint(): GeoPoint {
            return this.value.getGeoPoint();
        }

        getReference(): Reference {
            return this.value.getReference();
        }

        getBinaryReference(): BinaryReference {
            return this.value.getBinaryReference();
        }

        equals(o: api.Equitable): boolean {

            if (!api.ObjectHelper.iFrameSafeInstanceOf(o, Property)) {
                return false;
            }

            var other = <Property>o;

            if (!api.ObjectHelper.stringEquals(this.name, other.name)) {
                return false;
            }

            if (!api.ObjectHelper.numberEquals(this.index, other.index)) {
                return false;
            }

            if (!api.ObjectHelper.equals(this.value, other.value)) {
                return false;
            }

            return true;
        }

        copy(destinationPropertyArray: PropertyArray) {

            var destinationTree = destinationPropertyArray.getTree();

            var value: Value;
            if (this.value.isPropertySet() && this.value.isNotNull()) {
                var copiedPropertySet = this.value.getPropertySet().copy(destinationTree);
                value = new Value(copiedPropertySet, ValueTypes.DATA);
            }
            else {
                value = this.value;
            }

            var copy = Property.create().
                setName(this.name).
                setValue(value).
                setIndex(this.index).
                setArray(destinationPropertyArray).
                build();
            return copy;
        }

        onPropertyIndexChanged(listener: {(event: PropertyIndexChangedEvent): void;}) {
            this.propertyIndexChangedListeners.push(listener);
        }

        unPropertyIndexChanged(listener: {(event: PropertyIndexChangedEvent): void;}) {
            this.propertyIndexChangedListeners =
            this.propertyIndexChangedListeners.filter((curr) => (curr != listener));
        }

        private notifyPropertyIndexChangedEvent(previousIndex: number, newIndex: number) {
            var event = new PropertyIndexChangedEvent(this, previousIndex, newIndex);
            if (Property.debug) {
                console.debug("Property[" + this.getPath().toString() + "].notifyPropertyIndexChangedEvent: " + event.toString());
            }
            this.propertyIndexChangedListeners.forEach((listener) => listener(event));
        }

        onPropertyValueChanged(listener: {(event: PropertyValueChangedEvent): void;}) {
            this.propertyValueChangedListeners.push(listener);
        }

        unPropertyValueChanged(listener: {(event: PropertyValueChangedEvent): void;}) {
            this.propertyValueChangedListeners =
            this.propertyValueChangedListeners.filter((curr) => (curr != listener));
        }

        private notifyPropertyValueChangedEvent(previousValue: Value, newValue: Value) {
            var event = new PropertyValueChangedEvent(this, previousValue, newValue);
            if (Property.debug) {
                console.debug("Property[" + this.getPath().toString() + "].notifyPropertyValueChangedEvent: " + event.toString());
            }
            this.propertyValueChangedListeners.forEach((listener) => listener(event));
        }

        public static checkName(name: string) {
            if (name == null) {
                throw new Error("Property name cannot be null");
            }
            if (api.util.StringHelper.isBlank(name)) {
                throw new Error("Property name cannot be blank");
            }
            if (name.indexOf(".") >= 0) {
                throw new Error("Property name cannot contain .");
            }
            if (name.indexOf("[") >= 0 || name.indexOf("]") >= 0) {
                throw new Error("Property name cannot contain [ or ]");
            }
        }

        public static create(): PropertyBuilder {
            return new PropertyBuilder();
        }
    }

    export class PropertyBuilder {

        array: PropertyArray;

        name: string;

        index: number;

        value: Value;

        setArray(value: PropertyArray): PropertyBuilder {
            this.array = value;
            return this;
        }

        setName(value: string): PropertyBuilder {
            this.name = value;
            return this;
        }

        setIndex(value: number): PropertyBuilder {
            this.index = value;
            return this;
        }

        setValue(value: Value): PropertyBuilder {
            this.value = value;
            return this;
        }

        build(): Property {
            return new Property(this);
        }
    }
}

File: modules/core/core-api/src/main/java/com/enonic/xp/data/Property.java
package com.enonic.xp.data;

import java.time.Instant;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.LocalTime;
import java.util.Objects;

import com.enonic.xp.annotation.PublicApi;
import com.enonic.xp.util.BinaryReference;
import com.enonic.xp.util.GeoPoint;
import com.enonic.xp.util.Link;
import com.enonic.xp.util.Reference;

@PublicApi
public final class Property
{
    private final PropertyArray array;

    private final int index;

    private Value value;

    Property( final int index, final Value value, final PropertyArray array )
    {
        this.index = index;
        this.value = value;
        this.array = array;
        setPropertyOnPropertySetValue();
    }

    private void setPropertyOnPropertySetValue()
    {
        final Object valueObject = value.getObject();
        if (valueObject instanceof PropertySet)
        {
            final PropertySet propertySet = (PropertySet) valueObject;

            final PropertyTree tree = array.getParent().getTree();
            if ( tree != null && propertySet == tree.getRoot() )
            {
                throw new IllegalArgumentException( "Given PropertySet is the root PropertySet of the PropertyTree" );
            }
            propertySet.setProperty( this );
        }
    }

    public void setValue( final Value value )
    {
        this.array.checkType( value.getType() );
        this.value = value;
        this.setPropertyOnPropertySetValue();
    }

    public PropertySet getParent()
    {
        return array.getParent();
    }

    public String getName()
    {
        return array.getName();
    }

    public int getIndex()
    {
        return index;
    }

    public PropertyPath getPath()
    {
        if ( array.getParent().getProperty() != null )
        {
            return PropertyPath.from( this.array.getParent().getProperty().getPath(), PropertyPath.Element.from( array.getName(), index ) );
        }
        else
        {
            return PropertyPath.from( PropertyPath.Element.from( array.getName(), index ) );
        }
    }

    public ValueType getType()
    {
        return value.getType();
    }

    public Value getValue()
    {
        return value;
    }

    public boolean hasNullValue()
    {
        return value.isNull();
    }

    public boolean hasNotNullValue()
    {
        return !value.isNull();
    }

    public PropertySet getSet()
    {
        return value.asData();
    }

    public Object getObject()
    {
        return value.getObject();
    }

    public String getString()
    {
        return value.asString();
    }

    public Boolean getBoolean()
    {
        return value.asBoolean();
    }

    public Long getLong()
    {
        return value.asLong();
    }

    public Double getDouble()
    {
        return value.asDouble();
    }

    public GeoPoint getGeoPoint()
    {
        return value.asGeoPoint();
    }

    public Reference getReference()
    {
        return value.asReference();
    }

    public BinaryReference getBinaryReference()
    {
        return value.asBinaryReference();
    }

    public Link getLink()
    {
        return value.asLink();
    }

    public LocalDate getLocalDate()
    {
        return value.asLocalDate();
    }

    public LocalDateTime getLocalDateTime()
    {
        return value.asLocalDateTime();
    }

    public LocalTime getLocalTime()
    {
        return value.asLocalTime();
    }

    public Instant getInstant()
    {
        return value.asInstant();
    }

    public static void checkName( final String name )
    {
        if ( name == null )
        {
            throw new NullPointerException( "Property name cannot be null" );
        }
        if ( name.isBlank() )
        {
            throw new IllegalArgumentException( "Property name cannot be blank" );
        }
        if ( name.contains( "." ) )
        {
            throw new IllegalArgumentException( "Property name cannot contain ." );
        }
        if ( name.contains( "[" ) || name.contains( "]" ) )
        {
            throw new IllegalArgumentException( "Property name cannot contain [ or ]" );
        }
    }

    @Override
    public boolean equals( final Object o )
    {
        if ( this == o )
        {
            return true;
        }
        if ( !( o instanceof Property ) )
        {
            return false;
        }

        final Property property = (Property) o;

        if ( !Objects.equals( this.array.getName(), property.array.getName() ) )
        {
            return false;
        }
        if ( index != property.index )
        {
            return false;
        }
        return Objects.equals( value, property.value );
    }

    @Override
    public int hashCode()
    {
        return Objects.hash( array.getName(), index, value );
    }

    @Override
    public String toString()
    {
        return array.getName() + ": " + value;
    }

    public Property copyTo( final PropertySet destination )
    {
        return destination.addProperty( array.getName(), value.copy( destination.getTree() ) );
    }
}
